src/server/tiered_storage.cc

```diff
@@ -355,6 +356,68 @@ void TieredStorage::Free(PrimeIterator it, DbTableStats* stats) {
   stats->tiered_size -= len;
 }
 
+size_t TieredStorage::GetBinSize(size_t blob_len) {
+  CHECK(blob_len <= kMaxSmallBin);
+  unsigned bin_index = SmallToBin(blob_len);
+  return kSmallBins[bin_index];
+}
+
+void TieredStorage::Defrag(DbIndex db_index) {
+  // start scanning from a random position in the ref_cnt_ table.
+  auto refcnt_it = std::next(std::begin(page_refcnt_), rand() % page_refcnt_.size());
+
+  for (unsigned int i = 0; i < num_pages_to_scan_; ++i) {
+    // check the utilization of the page.
+    size_t bin_size = refcnt_it->second.second;
+    unsigned max_entries = NumEntriesInSmallBin(bin_size);
+    float bin_util = (float)(refcnt_it->second.first) / (float)max_entries;
+
+    // if bin is underutilized, start defragmentation.
+    if (bin_util < defrag_bin_util_threshold_) {
+      // 1. retrieve all the hash values stored in this page.
+      size_t hash_section_len = max_entries * kBytesPerHash;
+      std::vector<char> page(kBlockLen);
+      uint32_t offs_page = refcnt_it->first;
+      auto ec = Read(offs_page * kBlockLen, kBlockLen, &page[0]);
+      CHECK(!ec) << "Error when reading a page to be defragmented!";
+
+      for (unsigned int j = 0; j < max_entries; ++j) {
+        uint64_t key_hash = absl::little_endian::Load64(&page[j * kBytesPerHash]);
+        auto prime_it = db_slice_.GetDBTable(db_index)->prime.FindByHash(key_hash);
+
+        // if the key still exists, load the key into memory and reschedule offload
+        if (!prime_it.is_done()) {
+          PrimeValue* entry = &prime_it->second;
+          auto [offset, len] = entry->GetExternalSlice();
```

Comment:
call GetExternalSlice only you verify entry->IsExternal()

Reply:
you also need to verify that the offset is offset to this page , as we might have hash colisions of keys

Reply:
> call GetExternalSlice only you verify entry->IsExternal()

fixed.

Reply:
> you also need to verify that the offset is offset to this page , as we might have hash colisions of keys

that's a great point! i added a check. 

if collision happens, and the offset is not the right one, how do we keep retrieving the next entry?